include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,328 @@
+				uint8_t*	scratchMemory = nullptr;										//!< memory covering all regions used for temporary filling within computation of sum values
+				size_t	scratchMemoryByteSize = {};											//!< required byte size for entire scratch memory
+				bool normalizeImageByTotalABSValues = true;									//!< force normalizing by maximum absolute values
+
+				/*
+					layerCount - layer count used to execute the filter, not global layer count!
+					extent - extent of input image at chosen mip map level
+				*/
+
+				static inline size_t getRequiredScratchByteSize(size_t layerCount, asset::VkExtent3D extent)
+				{
+					size_t retval = extent.width * extent.height * extent.depth * decodeTypeByteSize * forcedScratchChannelAmount + (layerCount * decodeTypeByteSize * forcedScratchChannelAmount);
+					
+					return retval;
+				}
+
+				/*
+					Layer ID is relative to outBaseLayer in state
+				*/
+
+				enum E_SCALE_FACTOR
+				{
+					ESF_X,
+					ESF_Y,
+					ESF_COUNT
+				};
+
+				const float getAbsoluteLayerScaleValue(size_t layer, E_SCALE_FACTOR scaleFactor)
+				{
+					if (!scaleValuesPointer)
+					{
+						auto offset = layer * forcedScratchChannelAmount + scaleFactor == ESF_X ? 0 : 1;
+						return *(scaleValuesPointer + offset);
+					}
+					else
+						return 0; // or maybe assert?
+				}
+
+			protected:
+
+				float* scaleValuesPointer = nullptr;
+
+				std::vector<std::array<double, forcedScratchChannelAmount>> maxAbsLayerScaleValues; 								//!< scales gained by the filter (each layer handled) for derivative map shader usage
+		};
+
+	protected:
+		CNormalMapToDerivativeFilterBase() {}
+		virtual ~CNormalMapToDerivativeFilterBase() {}
+
+		static inline bool validate(CNormalMapToDerivativeStateBase* state)
+		{
+			if (!state)
+				return false;
+
+			if (!state->scratchMemory)
+				return false;
+
+			if (state->scratchMemoryByteSize == 0)
+				return false;
+
+			if (!impl::CSwizzleableAndDitherableFilterBase<false, false, Swizzle, IdentityDither>::validate(state))
+				return false;
+
+			return true;
+		}
+};
+
+//! Convert Normal Map to Derivative Normal Map
+/*
+	
+*/
+
+template<typename Swizzle = DefaultSwizzle>
+class CNormalMapToDerivativeFilter : public CMatchedSizeInOutImageFilterCommon, public CNormalMapToDerivativeFilterBase<Swizzle>
+{
+	public:
+		virtual ~CNormalMapToDerivativeFilter() {}
+
+		class CStateBase : public CMatchedSizeInOutImageFilterCommon::state_type, public CNormalMapToDerivativeFilterBase<Swizzle>::CNormalMapToDerivativeStateBase
+		{ 
+			public:
+				CStateBase() = default;
+				virtual ~CStateBase() = default;
+
+			private:
+
+				void setLayerScaleValuesOffset()
+				{
+					scaleValuesPointer = reinterpret_cast<float*>(scratchMemory) + (extent.width * extent.height * extent.depth * forcedScratchChannelAmount);
+				}
+
+				void resetLayerScaleValues()
+				{
+					memset(const_cast<float*>(scaleValuesPointer), 0, layerCount * forcedScratchChannelAmount * decodeTypeByteSize);
+				}
```

Comment:
not sure you need to explicitly reset if you're gonna overwrite all values anyway